From fa188266eb1ae8b7ff5d08b54b38029156678165 Mon Sep 17 00:00:00 2001
From: Carlos Leija <cileija@ti.com>
Date: Wed, 9 Mar 2011 15:14:05 -0600
Subject: OMAP4: Restoring PL310 Prefetch Offset Register after OFF

Current implementation was missing restoring the POR register
after hitting OFF

Change-Id: I5b23a01b05639c9f1cca2fe5a3e420924be3fcc0
Signed-off-by: Carlos Leija <cileija@ti.com>
---
 arch/arm/mach-omap2/include/mach/omap4-common.h |    3 +++
 arch/arm/mach-omap2/sleep44xx.S                 |   22 ++++++++++++++++++++++
 2 files changed, 25 insertions(+), 0 deletions(-)

diff --git a/arch/arm/mach-omap2/include/mach/omap4-common.h b/arch/arm/mach-omap2/include/mach/omap4-common.h
index f8ca41f..c29276f 100644
--- a/arch/arm/mach-omap2/include/mach/omap4-common.h
+++ b/arch/arm/mach-omap2/include/mach/omap4-common.h
@@ -67,8 +67,11 @@
 #define OMAP4_USBHOST_CLKSEL_UTMI_P2_EXT_P1_EXT		0x3
 
 /* L2 controller AUXCTRL value */
+#define OMAP4_L2X0_AUXCTL_VALUE_ES1			0x0e050000
 #define OMAP4_L2X0_AUXCTL_VALUE				0x1e470000
 
+#define A9_ES1_REV					0x410FC091
+
 #ifndef __ASSEMBLER__
 /*
  * wfi used in low power code. Directly opcode is used instead
diff --git a/arch/arm/mach-omap2/sleep44xx.S b/arch/arm/mach-omap2/sleep44xx.S
index 427caee..c38a934 100644
--- a/arch/arm/mach-omap2/sleep44xx.S
+++ b/arch/arm/mach-omap2/sleep44xx.S
@@ -57,6 +57,9 @@
 zero_params:
 .word 0
 
+POR_params:
+.word 1, 5
+
 /*
  * CPUx Wakeup Non-Secure Physical Address for
  * resume from OSWR/OFF
@@ -136,6 +139,25 @@ skip_enable_smp_bit:			@ GP device should execute from here
 	and	r0, #0x0f
 	cmp	r0, #1
 	beq	skip_l2en
+	/* Check ES revision */
+	mrc	p15, 0, r0, c0, c0, 0
+	ldr	r1, =A9_ES1_REV
+	cmp	r0, r1
+	bne	check_por
+	ldr	r0, =OMAP4_L2X0_AUXCTL_VALUE_ES1
+	ldr     r12, =0x109		@ Setup L2 AUXCTL value
+	dsb
+	smc     #0
+	b	skip_l2en
+check_por:
+	ldr	r0, =OMAP44XX_SAR_RAM_BASE  @ Check DEVICE type
+	ldr	r1, [r0, #OMAP_TYPE_OFFSET]
+	cmp	r1, #0x1                   @ Check for HS device
+	bne	skip_por
+	ldr	r0, =PPA_SERVICE_PL310_POR @ Setup PPA HAL call
+	adr	r3, POR_params
+	LM_CALL_PPA_SERVICE_PA
+skip_por:
 	ldr	r0, =OMAP4_L2X0_AUXCTL_VALUE
 	ldr     r12, =0x109		@ Setup L2 AUXCTL value
 	dsb
-- 
1.6.0.4

